### PR TITLE
Firewoodize merkle db Part 1: Make Views ReadOnly

### PR DESCRIPTION
**Pull request by dboehm-avalabs**
Remove the ability to insert/remove key values from merkledb directly.  Can make changes by passing them in as a batch when creating a new view.

### CODE REVIEW

File: x/merkledb/db.go [hunk 1]
```diff
@@ -634,35 +654,31 @@ func (db *merkleDB) GetChangeProof(
...
+// Assumes [db.commitLock] isn't held.
+func (db *merkleDB) NewView(batchOps []database.BatchOp) (TrieView, error) {
+	db.commitLock.RLock()
+	defer db.commitLock.RUnlock()
+
+	newView, err := db.newUntrackedView(batchOps)
+	if err != nil {
+		return nil, err
+	}
+	db.childViews = append(db.childViews, newView)
```
Comment (darioush) [inline x/merkledb/db.go:666]: doesn't taking RLock here make it possible for 2 different appends to happen concurrently?

File: x/merkledb/trie_test.go [hunk 2]
```diff
@@ -792,13 +789,13 @@ func Test_Trie_MultipleStates(t *testing.T) {
...
 			// Populate initial set of keys
-			root, err := db.NewView()
+			root, err := db.NewView(nil)
 			require.NoError(err)
 			kv := [][]byte{}
 			for i := 0; i < initialSet; i++ {
 				k := []byte(strconv.Itoa(i))
 				kv = append(kv, k)
-				require.NoError(root.Insert(context.Background(), k, hashing.ComputeHash256(k)))
+				require.NoError(root.(*trieView).insert(k, hashing.ComputeHash256(k)))
```
Comment (darioush) [inline x/merkledb/trie_test.go:798]: nit: seems we can do this without the cast, building an `ops`  array then calling db.NewView

File: x/merkledb/proof.go [hunk 3]
```diff
@@ -845,7 +847,7 @@ func addPathInfo(
 	return nil
 }
 
-func getEmptyTrieView(ctx context.Context) (*trieView, error) {
+func getTrieView(ctx context.Context, ops []database.BatchOp) (*trieView, error) {
```
Comment (danlaine) [inline x/merkledb/proof.go:850]: Trying to think of a better name for this. `newStandaloneTrieView` or something? Idk.

Comment (danlaine) [reply x/merkledb/proof.go:850]: Maybe a comment would suffice.

File: x/merkledb/trie.go [hunk 4]
```diff
@@ -53,17 +53,8 @@ type ReadOnlyTrie interface {
...
-	Remove(ctx context.Context, key []byte) error
-
-	// NewPreallocatedView returns a new view on top of this Trie with space allocated for changes
-	NewPreallocatedView(estimatedChanges int) (TrieView, error)
-
-	// NewView returns a new view on top of this Trie
-	NewView() (TrieView, error)
-
-	// Insert a key/value pair into the Trie
-	Insert(ctx context.Context, key, value []byte) error
```
Comment (danlaine) [inline x/merkledb/trie.go:66]: :heart:

File: x/merkledb/trie_test.go [hunk 5]
```diff
@@ -281,56 +271,34 @@ func Test_Trie_InsertAndRetrieve(t *testing.T) {
...
-	require.NoError(trie.Insert(context.Background(), []byte("key"), []byte("value")))
+	require.NoError(dbTrie.Put([]byte("key"), []byte("value")))
 
-	value, err = getNodeValue(trie, "key")
+	value, err = getNodeValue(dbTrie, "key")
 	require.NoError(err)
 	require.Equal([]byte("value"), value)
 }
 
-func Test_Trie_Overwrite(t *testing.T) {
```
Comment (danlaine) [inline x/merkledb/trie_test.go:298]: How come this got deleted?

File: x/merkledb/trieview.go [hunk 6]
```diff
@@ -789,14 +773,6 @@ func (t *trieView) getValue(key path, lock bool) ([]byte, error) {
...
-func (t *trieView) Insert(_ context.Context, key []byte, value []byte) error {
-	t.lock.Lock()
-	defer t.lock.Unlock()
-
-	return t.insert(key, value)
-}
-
 // Assumes [t.lock] is held.
 // Assumes [t.validityTrackingLock] isn't held.
 func (t *trieView) insert(key []byte, value []byte) error {
```
Comment (danlaine) [inline x/merkledb/trieview.go:778]: Can we add a comment to clarify the difference between `insert` and `insertIntoTrie`?

File: x/merkledb/trieview.go [hunk 7]
```diff
@@ -845,11 +816,13 @@ func (t *trieView) remove(key []byte) error {
 
 	// the trie has been changed, so invalidate all children and remove them from tracking
 	t.invalidateChildren()
-
-	if err := t.recordValueChange(newPath(key), maybe.Nothing[[]byte]()); err != nil {
+	path := newPath(key)
+	if err := t.recordValueChange(path, maybe.Nothing[[]byte]()); err != nil {
+		return err
+	}
+	if err := t.removeFromTrie(path); err != nil {
```
Comment (danlaine) [inline x/merkledb/trieview.go:826]: This is only called here...should we keep this as a separate function?

**Review by danlaine — APPROVED** [review]
Looks OK to me but two things I want to flag:

1. We still need to update the readme 
2. The code is in a kind of intermediate state since we've made views read only, but still have a lot of code (mostly locking) that helped to handle views being editable. Should we wait until subsequent PRs cleaning up that code have been written/merged into this one before merging this into dev?

File: x/merkledb/trie_test.go (hunk 5 above)
Comment (dboehm-avalabs) [reply x/merkledb/trie_test.go:298]: you can't overwrite if it is only writeable on creation

Comment (danlaine) [reply x/merkledb/trie_test.go:298]: It's possible to call `insert` multiple times with the same key. Namely, if the batch passed into `NewView` contains two `BatchOp` with the same key.

**Comment by dboehm-avalabs** [comment]
> Looks OK to me but two things I want to flag:
> 
> 1. We still need to update the readme
> 2. The code is in a kind of intermediate state since we've made views read only, but still have a lot of code (mostly locking) that helped to handle views being editable. Should we wait until subsequent PRs cleaning up that code have been written/merged into this one before merging this into dev?

I think we risk having an enormous, unreviewable mess if we try to merge everything at once.  I think iterative cleanup might be best.

**Pull request merged**